src/thing_creature.c

```diff
@@ -1242,18 +1255,204 @@ void process_thing_spell_teleport_effects(struct Thing *thing, struct CastedSpel
+                            {
+                                if (i > BATTLES_COUNT)
+                                {
+                                    i = 1;
+                                    battleid = 1;
+                                }
+                                count++;
+                                struct CreatureBattle* battle = creature_battle_get(i);
+                                if ( (battle->fighters_num != 0) && (battle_with_creature_of_player(thing->owner, i)) )
+                                {
+                                    struct Thing* tng = thing_get(battle->first_creatr);
+                                    TRACE_THING(tng);
+                                    if (creature_can_navigate_to_with_storage(thing, &tng->mappos, NavRtF_Default))
+                                    {
+                                        pos.x.val = tng->mappos.x.val;
+                                        pos.y.val = tng->mappos.y.val;
+                                        battleid = i + 1;
+                                        break;
+                                    }
+                                }
+                                if (count >= BATTLES_COUNT)
+                                {
+                                    battleid = 1;
+                                    break;
+                                }
+                                if (i >= BATTLES_COUNT)
+                                {
+                                    i = 0;
+                                    battleid = 1;
+                                    continue;
+                                }
+                            }
+                        }
+                        else
+                        {
+                            allowed = false;
+                        }
+                        break;
+                    }
+                    case 17:
+                    {
+                        room = room_get(cctrl->last_work_room_id);
+                        allowed = (!room_is_invalid(room));
+                        break;
+                    }
+                    case 18:
+                    {
+                        struct Coord3d cta_pos;
+                        cta_pos.x.val = subtile_coord_center(dungeon->cta_stl_x);
+                        cta_pos.y.val = subtile_coord_center(dungeon->cta_stl_y);
+                        cta_pos.z.val = subtile_coord(1,0);
+                        if (creature_can_navigate_to_with_storage(thing, &cta_pos, NavRtF_Default))
+                        {
+                        pos = cta_pos;
+                        }
+                        else
+                        {
+                            allowed = false;
+                        }
+                        break;
+                    }
+                    case 19:
+                    {
+                        desttng = find_hero_door_hero_can_navigate_to(thing);
```

Comment:
I would actually remove case 19, the hero door. It's too powerful an ability, as it allows you to actually discover the hero portals by teleporting to them all.

Reply:
It only works if the creature can get there normally. That's how find_hero_door_hero_can_navigate_to works.

Reply:
Well, 'being able to get there' is something else than 'having explored it'. For rooms this is not the case, for CtA, Annoyed Creature, Fight, you all have it explored.

Hero gates behind doors, across lava, or even simply in unexplored areas beyond heroes on guard posts, are not explored yet.